[Merged by Bors] - feat(algebra/order/to_interval_mod): add circular_order instance for add_circle

eric-wieser opened the pull request:
This also provides us with the same instance for `real.angle`.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
-->
- [x] depends on: #17739
- [x] depends on: #17741
- [x] depends on: #17933
- [x] depends on: #18912 
- [x] depends on: #18942
- [x] depends on: #18955
- [x] depends on: #18958

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Note the only modification to a synchronized file is removing a TODO comment.

eric-wieser committed:
chore(algebra/order/to_interval_mod): add `equiv` versions extracted from `add_circle`

eric-wieser committed:
wip

eric-wieser committed:
wip

eric-wieser committed:
feat(algebra/order/to_interval_mod): lemmas about changing the base of the period

eric-wieser committed:
Merge branch 'eric-wieser/to_Ico_div-lemmas' into eric-wieser/to_Ico_mod-btw

eric-wieser committed:
wip

eric-wieser committed:
extract sorries to standalone lemmas

eric-wieser committed:
add the punchline

eric-wieser committed:
some non-progress

eric-wieser committed:
some more helper lemmas

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/to_Ico_mod-btw

eric-wieser committed:
wip

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/to_Ico_mod-btw

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/to_Ico_mod-btw

eric-wieser committed:
one sorry in terms of another

eric-wieser committed:
feat(algebra/order/to_interval_mod): relation between `to_Ixx_{mod/div}` definitions

`to_Ico_div` and `to_Ioc_div` differ by at most 1, and the modulo versions differ by at most `b`.

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/to_Ico_mod-lemmas

eric-wieser committed:
use new lemmas

eric-wieser committed:
golf

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/to_Ico_mod-lemmas

eric-wieser committed:
wip

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/to_Ico_mod-lemmas

eric-wieser committed:
fix some errors

eric-wieser committed:
more fixes

eric-wieser committed:
fix

eric-wieser committed:
cleanup

eric-wieser committed:
one more lemma

eric-wieser committed:
Merge remote-tracking branch 'origin/eric-wieser/to_Ico_mod-lemmas' into eric-wieser/to_Ico_mod-btw

eric-wieser committed:
adjust to new API

eric-wieser committed:
line wrap comment

eric-wieser committed:
wip

eric-wieser committed:
golf and add a few more lemmas

eric-wieser committed:
remove another redundant lemma

eric-wieser committed:
Merge commit 'a993e7ee843ff31f2d474197483d2f6d6610addd' into eric-wieser/to_Ico_mod-lemmas

eric-wieser committed:
Merge branch 'eric-wieser/to_Ico_mod-lemmas' into eric-wieser/to_Ico_mod-btw

YaelDillies reviewed (commented):
Are you feeling like we're missing a constructor for `circular_order`?

YaelDillies committed:
push antisymmetry sorry back

eric-wieser committed:
wip

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/to_Ico_mod-btw

eric-wieser committed:
move some results so that a sorry can be solved

eric-wieser committed:
fix long lines

collares committed:
break sorries into smaller sorries (feel free to revert)

collares committed:
sorry-free, still very ugly (feel free to revert)

collares committed:
slightly less ugly (feel free to revert)

collares committed:
slightly less ugly (feel free to revert)

collares committed:
slightly less ugly (feel free to revert)

eric-wieser committed:
refactor(algebra/order/to_interval_mod): Negate the definition of `mem_Ioo_mod`

This replaces `mem_Ioo_mod hp a b` with `¬Imodeq p a b`, where `Imodeq` stands for `interval mod equal`.
This is more consistent with `int.modeq`, `nat.modeq`, and `smodeq`.

There's still some duplication here, but at least these four ideas are now conceptually aligned.

This remove any lemmas of the form `¬Imodeq p a b ↔ _ ≠ _` as these are now trivial consequences of the `Imodeq p a b ↔ _ = _` versions,

eric-wieser committed:
Apply suggestions from code review

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>
Co-authored-by: Junyan Xu <junyanxumath@gmail.com>

eric-wieser committed:
golf

eric-wieser committed:
blank line

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/to_Ico_mod-btw

YaelDillies committed:
review API

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/to_Ico_mod-btw

eric-wieser committed:
move some lemmas

eric-wieser committed:
cleanup

eric-wieser committed:
move lemmas again

eric-wieser committed:
fix name

eric-wieser committed:
remove underscores

eric-wieser commented on src/algebra/order/to_interval_mod.lean:
```diff
@@ -181,11 +182,27 @@ by { rw [to_Ioc_mod_eq_iff hp, set.right_mem_Ioc], exact ⟨lt_add_of_pos_right
+  rw [sub_smul, ←sub_add, add_right_comm],
+  simpa using sub_to_Ico_div_zsmul_mem_Ico hp a b,
+end
+
```
I've split out these lemmas to #18942

YaelDillies committed:
finish

YaelDillies committed:
fixes

YaelDillies committed:
fix topology.instances.add_circle

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/to_Ico_mod-btw

eric-wieser committed:
fix merge errors

eric-wieser committed:
feat(algebra/order/to_interval_mod): notation for `add_comm_group.modeq`

eric-wieser committed:
Merge branch 'eric-wieser/modeq-notation' into eric-wieser/to_Ico_mod-btw

YaelDillies committed:
expand TODO

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into redef_add_comm_group_modeq

YaelDillies committed:
update module docs, move div_mul_cancel'''

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/to_Ico_mod-btw

YaelDillies committed:
nsmul lemmas

YaelDillies committed:
move int.cast_mul'

YaelDillies committed:
use the new notation

YaelDillies committed:
Merge remote-tracking branch 'origin/redef_add_comm_group_modeq' into eric-wieser/to_Ico_mod-btw

collares committed:
Merge remote-tracking branch 'origin/staging' into eric-wieser/to_Ico_mod-btw

collares committed:
spacing

eric-wieser committed:
golf

eric-wieser committed:
reduce diff

collares committed:
golf

collares committed:
remove redundant parentheses

collares committed:
small reordering

collares committed:
golf

collares committed:
golf

collares committed:
golf

collares committed:
no more golfing for me

collares committed:
prettify to_Ixx_mod_trans
